tests/controllers/test_registry_instrument.py

```diff
@@ -0,0 +1,196 @@
+        session = MagicMock()
+
+        mock_registry_instrument = MagicMock()
+        mock_registry_instrument.to_dict.return_value = {"dummy_key": "dummy_value_1"}
+        mock_query.return_value.first.return_value = mock_registry_instrument
+
+        controller = RegistryInstrument()
+        exercise_id = "3ff59b73-7f15-406f-9e4d-7f00b41e85ce"
+        formtype = "0001"
+
+        result = controller.get_registry_instrument_by_exercise_id_and_formtype.__wrapped__(
+            controller, exercise_id, formtype, session
+        )
+
+        mock_query.assert_called_once()
+        mock_registry_instrument.to_dict.assert_called_once()
+        self.assertEqual(result, {"dummy_key": "dummy_value_1"})
+
+    @patch("application.controllers.registry_instrument.query_registry_instrument_by_exercise_id_and_formtype")
+    def test_get_registry_instrument_by_exercise_id_and_formtype_returns_none(self, mock_query):
+        session = MagicMock()
+
+        mock_registry_instrument = MagicMock()
+        mock_query.return_value.first.return_value = None
+
+        controller = RegistryInstrument()
+        exercise_id = "3ff59b73-7f15-406f-9e4d-7f00b41e85ce"
+        formtype = "0001"
+
+        result = controller.get_registry_instrument_by_exercise_id_and_formtype.__wrapped__(
+            controller, exercise_id, formtype, session
+        )
+
+        mock_query.assert_called_once()
+        mock_registry_instrument.to_dict.assert_not_called()
+        self.assertEqual(result, None)
+
+    @patch("application.controllers.registry_instrument.query_registry_instrument_by_exercise_id_and_formtype")
+    def test_save_new_registry_instrument_for_exercise_id_and_formtype(self, mock_query):
+        session = MagicMock()
+
+        # Mock a registry instrument as NOT in the database
+        mock_query.return_value.first.return_value = None
+
+        # Set up the validated data from the posted payload
+        ci_version = 99
+        form_type = "0002"
+        exercise_id = "3ff59b73-7f15-406f-9e4d-7f00b41e85ce"
+        guid = "678a583a-87f6-4c57-9f5b-12c5ced30c1e"
+        published_at = "2028-01-30T12:00:00"
+        survey_id = "0b1f8376-28e9-4884-bea5-acf9d709464e"
+        instrument_id = "77c1e406-716a-488d-bfc9-b5b988fbaccf"
+
+        controller = RegistryInstrument()
+
+        with patch(
+            "application.controllers.registry_instrument.RegistryInstrumentModel"
+        ) as mock_registry_instrument_model:
+            result = controller.save_registry_instrument_for_exercise_id_and_formtype.__wrapped__(
+                controller, survey_id, exercise_id, instrument_id, form_type, ci_version, published_at, guid, session
+            )
+
+        # Assert that a new instrument was created and saved
+        mock_query.assert_called_once()
+        mock_registry_instrument_model.assert_called_once()
+        self.assertEqual(mock_registry_instrument_model.return_value.ci_version, ci_version)
```

Comment:
Feels like this could be done a bit neater, i.e called to dict and then compare that dict to what you expect

Reply:
I looked at that originally but had real trouble with the mock object return value giving me the right thing to assert against.

Or do you mean something like this?

```
expected = {
    "ci_version": ci_version,
    "guid": guid,
    "published_at": published_at,
...
...
blah...
}

actual = {
    "ci_version": mock_registry_instrument_model.return_value.ci_version,
    "guid": mock_registry_instrument_model.return_value.guid,
    "published_at": mock_registry_instrument_model.return_value.published_at,
...
...
blah...
}

self.assertEqual(actual, expected)
```

Reply:
Yep, that would work